File: src/main/java/net/draconia/askaround/domain/AssetImage.java
package net.draconia.askaround.domain;

import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.Serializable;

import javax.imageio.ImageIO;

import net.draconia.askaround.util.ImageSerializer;

import org.apache.commons.codec.binary.Base64;
import org.codehaus.jackson.annotate.JsonIgnore;
import org.codehaus.jackson.map.annotate.JsonSerialize;

public class AssetImage implements Serializable
{
	private static final long serialVersionUID = -1032171751975532303L;
	
	private Integer miHeight, miId, miWidth;
	private String msDescription, msName;
	private Image mObjData;
	
	public AssetImage()
	{ }
	
	public AssetImage(final Image objData)
	{
		setData(objData);
	}
	
	public AssetImage(final Integer iWidth, final Integer iHeight)
	{
		setWidth(iWidth);
		setHeight(iHeight);
	}
	
	public AssetImage(final Image objData, final String sName)
	{
		setData(objData);
		setName(sName);
	}
	
	public AssetImage(final Image objData, final Integer iWidth, final Integer iHeight)
	{
		setData(objData);
		setWidth(iWidth);
		setHeight(iHeight);
	}
	
	public AssetImage(final Image objData, final Integer iWidth, final Integer iHeight, final String sName)
	{
		setData(objData);
		setWidth(iWidth);
		setHeight(iHeight);
		setName(sName);
	}
	
	public AssetImage(final Image objData, final Integer iWidth, final Integer iHeight, final String sName, final String sDescription)
	{
		setData(objData);
		setWidth(iWidth);
		setHeight(iHeight);
		setName(sName);
		setDescription(sDescription);
	}
	
	public AssetImage(final Integer iId, final Image objData, final Integer iWidth, final Integer iHeight, final String sName, final String sDescription)
	{
		setId(iId);
		setData(objData);
		setWidth(iWidth);
		setHeight(iHeight);
		setName(sName);
		setDescription(sDescription);
	}
	
	public boolean equals(final Object objOther)
	{
		if((objOther instanceof AssetImage) && (objOther != null))
			{
			AssetImage objImage = ((AssetImage)(objOther));
			
			return(		(getId() == objImage.getId())
					&&	(getData().equals(objImage.getData()))
					&&	(getWidth() == objImage.getWidth())
					&&	(getHeight() == objImage.getHeight())
					&&	(getName().equals(objImage.getName()))
					&&	(getDescription().equals(objImage.getDescription())));
			}
		else
			return(false);
	}
	
	@JsonIgnore
	public String getBase64() throws IOException
	{
		ByteArrayOutputStream objStream = new ByteArrayOutputStream();
		
		ImageIO.write(((BufferedImage)(getData())), "jpg", objStream);
		
		objStream.flush();
		
		return(Base64.encodeBase64String(objStream.toByteArray()));
	}
	
	/**
	 * getData - Returns the image for this thumbnail
	 * @return image/thumbnail in the Asset record
	 */
	// The annotation permits the Jackson library to encode the image into the JSON
	@JsonSerialize(using = ImageSerializer.class)
	public Image getData()
	{
		if(mObjData == null)
			mObjData = new BufferedImage(32, 32, BufferedImage.TYPE_4BYTE_ABGR);
		
		return(mObjData);
	}
	
	public String getDescription()
	{
		if(msDescription == null)
			msDescription = "";
		
		return(msDescription);
	}
	
	public int getHeight()
	{
		if(miHeight == null)
			miHeight = ((BufferedImage)(getData())).getHeight();
		
		return(miHeight);
	}
	
	public int getId()
	{
		if(miId == null)
			miId = 0;
		
		return(miId);
	}
	
	public String getName()
	{
		if(msName == null)
			msName = "";
		
		return(msName);
	}
	
	public int getWidth()
	{
		if(miWidth == null)
			miWidth = ((BufferedImage)(getData())).getWidth();
		
		return(miWidth);
	}
	
	public void setData(final Image objData)
	{
		if(objData == null)
			mObjData = new BufferedImage(32, 32, BufferedImage.TYPE_4BYTE_ABGR);
		else
			mObjData = objData;
	}
	
	public void setDescription(final String sDescription)
	{
		if(sDescription == null)
			msDescription = "";
		else
			msDescription = sDescription;
	}
	
	public void setHeight(final Integer iHeight)
	{
		if(iHeight == null)
			miHeight = 32;
		else
			miHeight = iHeight;
	}
	
	public void setId(final Integer iId)
	{
		if(iId == null)
			miId = 0;
		else
			miId = iId;
	}
	
	public void setName(final String sName)
	{
		if(sName == null)
			msName = "";
		else
			msName = sName;
	}
	
	public void setWidth(final Integer iWidth)
	{
		if(iWidth == null)
			miWidth = 32;
		else
			miWidth = iWidth;
	}
}

File: src/main/webapp/scripts/models/assetimage.js
function AssetImage(iId, objData, iWidth, iHeight, sName, sDescription)
{
	this.mbChanged						=	undefined;
	this.mObjData						=	objData;
	this.msDescription					=	Common.defaultIfUndefined(sDescription, "");
	this.miHeight 						= 	Common.defaultIfUndefined(iHeight, 32);
	this.miId							=	Common.defaultIfUndefined(iId, 0);
	this.msName							=	Common.defaultIfUndefined(sName, "");
	this.mArrObjObservers				=	[];
	this.miWidth						=	Common.defaultIfUndefined(iWidth, 32);
};

AssetImage.prototype.addObserver		=	function(objObserver)
{
	if(objObserver != undefined)
		{
		if(this.mArrObjObservers == undefined)
			this.mArrObjObservers = [];
		
		this.mArrObjObservers.push(objObserver);
		}
	
	return(this);
};

AssetImage.copyFromJSON				=	function(objJSON, objAssetImage)
{
	if(objAssetImage == undefined)
		objAssetImage = new AssetImage();
	
	for(prop in objJSON)
		objAssetImage["set" + prop.substring(0, 1).toUpperCase() + prop.substring(1)].call(objAssetImage, eval("objJSON." + prop));
	
	return(objAssetImage);
};

AssetImage.prototype.getData			=	function()
{
	return(this.mObjData);
};

AssetImage.prototype.getDescription		=	function()
{
	if(this.msDescription == undefined)
		this.msDescription = "";
	
	return(this.msDescription);
};

AssetImage.prototype.getHeight			=	function()
{
	if(this.miHeight == undefined)
		this.miHeight = 32;
	
	return(this.miHeight);
};

AssetImage.prototype.getId				=	function()
{
	if(this.miId == undefined)
		this.miId = 0;
	
	return(this.miId);
};

AssetImage.prototype.getName			=	function()
{
	if(this.msName == undefined)
		this.msName = "";
	
	return(this.msName);
};

AssetImage.prototype.getWidth			=	function()
{
	if(this.miWidth == undefined)
		this.miWidth = 32;
	
	return(this.miWidth);
};

AssetImage.prototype.hasChanged			=	function()
{
	if(this.mbChanged == undefined)
		this.mbChanged = false;
	
	return(this.mbChanged);
};

AssetImage.prototype.notifyObservers	=	function(objArgument)
{
	this.mbChanged = true;
	
	for(var iLength = this.mArrObjObservers.length, iLoop = 0; iLoop < iLength; iLoop++)
		this.mArrObjObservers[iLoop].call(this, objArgument);
	
	this.mbChanged = false;
};

AssetImage.prototype.packageToJSON		=	function()
{
	return(	{	id:				this.getId()
			,	data:			this.getData()
			,	name:			this.getName()
			,	description:	this.getDescription()
			,	height:			this.getHeight()
			,	width:			this.getWidth()
			});
};

AssetImage.prototype.setData			=	function(objData)
{
	this.mObjData = objData;
	
	this.notifyObservers();
	
	return(this);
};

AssetImage.prototype.setDescription		=	function(sDescription)
{
	if(sDescription == undefined)
		this.msDescription = "";
	else
		this.msDescription = sDescription;
	
	this.notifyObservers();
	
	return(this);
};

AssetImage.prototype.setHeight			=	function(iHeight)
{
	if(iHeight == undefined)
		this.miHeight = 32;
	else
		this.miHeight = iHeight;
	
	this.notifyObservers();
	
	return(this);
};

AssetImage.prototype.setId				=	function(iId)
{
	if(iId == undefined)
		this.miId = 0;
	else
		this.miId = iId;
	
	this.notifyObservers();
	
	return(this);
};

AssetImage.prototype.setName			=	function(sName)
{
	if(sName == undefined)
		this.msName = "";
	else
		this.msName = sName;
	
	this.notifyObservers();
	
	return(this);
};

AssetImage.prototype.setWidth			=	function(iWidth)
{
	if(iWidth == undefined)
		this.miWidth = 32;
	else
		this.miWidth = iWidth;
	
	this.notifyObservers();
	
	return(this);
};